go/cmd/dolt/commands/push.go

```diff
@@ -85,95 +95,236 @@ func (cmd PushCmd) Exec(ctx context.Context, commandStr string, args []string, d
+				"upstream, see 'push.autoSetupRemote' in 'dolt config --help'.")
+		}
+		if strings.Contains(text, "upstream branch already set for") {
+			// success
+			return nil
+		}
+		if strings.Contains(text, doltdb.ErrIsAhead.Error()) ||
+			strings.Contains(text, actions.ErrCantFF.Error()) {
+			srcRef, destRef, err := getTrackingRefs(branchName, targetRemote)
 			if err != nil {
-				verr = errhand.BuildDError("fatal: The current branch could not be identified").AddCause(err).Build()
-			} else {
-				remoteName := "<remote>"
-				if defRemote, verr := env.GetDefaultRemote(dEnv.RepoStateReader()); verr == nil {
-					remoteName = defRemote.Name
-				}
-				verr = errhand.BuildDError("fatal: The current branch " + currentBranch.GetPath() + " has no upstream branch.\n" +
-					"To push the current branch and set the remote as upstream, use\n" +
-					"\tdolt push --set-upstream " + remoteName + " " + currentBranch.GetPath() + "\n" +
-					"To have this happen automatically for branches without a tracking\n" +
-					"upstream, see 'push.autoSetupRemote' in 'dolt config --help'.").Build()
+				return fmt.Errorf("failed to get tracking ref info: %w", err)
 			}
 
-		case env.ErrInvalidSetUpstreamArgs:
-			verr = errhand.BuildDError("error: --set-upstream requires <remote> and <refspec> params.").SetPrintUsage().Build()
-		default:
-			verr = errhand.VerboseErrorFromError(err)
+			cli.Printf("To %s\n", targetRemote.Url)
+			cli.Printf("! [rejected]          %s -> %s (non-fast-forward)\n", srcRef, destRef)
+			cli.Printf("error: failed to push some refs to '%s'\n", targetRemote.Url)
+			cli.Println("hint: Updates were rejected because the tip of your current branch is behind")
+			cli.Println("hint: its remote counterpart. Integrate the remote changes (e.g.")
+			cli.Println("hint: 'dolt pull ...') before pushing again.")
 		}
-		return HandleVErrAndExitCode(verr, usage)
+		return err
 	}
 
-	remoteDB, err := opts.Remote.GetRemoteDB(ctx, dEnv.DoltDB.ValueReadWriter().Format(), dEnv)
+	postPushRemoteBranches, err := getRemoteBranches(queryist, sqlCtx)
 	if err != nil {
-		err = actions.HandleInitRemoteStorageClientErr(opts.Remote.Name, opts.Remote.Url, err)
-		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+		return fmt.Errorf("failed to get post-push remote branches: %w", err)
 	}
 
-	tmpDir, err := dEnv.TempTableFilesDir()
-	if err != nil {
-		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+	changesMade := getChangesMade(remoteBranches, postPushRemoteBranches)
+	if !changesMade {
+		cli.Println("Everything up-to-date")
 	}
-	var verr errhand.VerboseError
-	err = actions.DoPush(ctx, dEnv.RepoStateReader(), dEnv.RepoStateWriter(), dEnv.DoltDB, remoteDB, tmpDir, opts, buildProgStarter(defaultLanguage), stopProgFuncs)
+
+	return nil
+}
+
+func getChangesMade(remoteBranches map[string]remoteBranchInfo, postPushRemoteBranches map[string]remoteBranchInfo) bool {
```

Comment:
Docs please.

I would rename the arguments to be prePushBranches and postPushBranches. And, I'm pretty sure you just want to compare them for equality which doesn't require all the logic below.

Reply:
Sure thing. Will try a basic equality comparison, hopefully that'll be enough.